worker/deps/libwebrtc/libwebrtc/modules/congestion_controller/goog_cc/robust_throughput_estimator.cc

```diff
@@ -0,0 +1,189 @@
+/*
+ *  Copyright (c) 2019 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
+
+#include "modules/congestion_controller/goog_cc/robust_throughput_estimator.h"
+
+#include <stddef.h>
+
+#include <algorithm>
+#include <utility>
+
+#include "api/units/data_rate.h"
+#include "api/units/data_size.h"
+#include "api/units/time_delta.h"
+#include "api/units/timestamp.h"
+// #include "rtc_base/checks.h"
+
+namespace webrtc {
+
+RobustThroughputEstimator::RobustThroughputEstimator(
+    const RobustThroughputEstimatorSettings& settings)
+    : settings_(settings),
+      latest_discarded_send_time_(Timestamp::MinusInfinity()) {
+  // RTC_DCHECK(settings.enabled);
+}
+
+RobustThroughputEstimator::~RobustThroughputEstimator() {}
+
+bool RobustThroughputEstimator::FirstPacketOutsideWindow() {
+  if (window_.empty())
+    return false;
+  if (window_.size() > settings_.max_window_packets)
+    return true;
+  TimeDelta current_window_duration =
+      window_.back().receive_time - window_.front().receive_time;
+  if (current_window_duration > settings_.max_window_duration)
+    return true;
+  if (window_.size() > settings_.window_packets &&
+      current_window_duration > settings_.min_window_duration) {
+    return true;
+  }
+  return false;
+}
+
+void RobustThroughputEstimator::IncomingPacketFeedbackVector(
+    const std::vector<PacketResult>& packet_feedback_vector) {
+  // RTC_DCHECK(std::is_sorted(packet_feedback_vector.begin(),
+  //                          packet_feedback_vector.end(),
+  //                          PacketResult::ReceiveTimeOrder()));
+  for (const auto& packet : packet_feedback_vector) {
+    // Ignore packets without valid send or receive times.
+    // (This should not happen in production since lost packets are filtered
+    // out before passing the feedback vector to the throughput estimator.
+    // However, explicitly handling this case makes the estimator more robust
+    // and avoids a hard-to-detect bad state.)
+    if (packet.receive_time.IsInfinite() ||
```

Comment:
in case it happens, a warn log is helpful becasue it means somewhere else in the code there is a problem!

Reply:
Robust throughput estimator is meant to be a better alternative to acknowledged bitrate estimator. But is disabled by default in favour of acknowledged bitrate estimator. Controlled by filed trial. But I cant see any ongoing Chrome expiriment for it, and have not tested to see the difference.

Reply:
Read this for a reference:
https://bugs.chromium.org/p/webrtc/issues/detail?id=10274
https://bugs.chromium.org/p/webrtc/issues/detail?id=13402

Reply:
Right! Only suggested a warn log in case it is invoked especially with a packet does not have a receive_time. If this part of the code is not invoked for the new estimator, should it be marked as deprecated?